### ISSUE TITLE
zstd: Not enough Memory when using threads >1

### ISSUE THREAD

**Issue by deadair3210** [nicoboss/nsz#37]
No matter the source file size, if I try to use any more than 1 thread for compressing nsp to nsz, the script dies with a not enough memory error, I have 24GB of RAM, surely its a error, no?

**Comment by nicoboss** [comment]
I just tried Noire_0100830004FB6800_[v131072] with --threads 3 and --level 0 - it never went over 300 MB RAM usage on latest commit. So I thought it might be an issue with nsz_v2.1.1_win64_portable so I compressed SNES_Online_01008D300C50C000_[v0] with --threads 3 and the default compression level with it but RAM usage still was normal. Just to make sure I tried --threads 3000 but RAM usage was still normal. Because I saw nothing special, I tried the same using block compression instead of solid compression but with no results. Only block compression with an unreasonably high amount of threads filled up my RAM.

To summarize I tried and wasn’t able to reproduce your issue. Please tell me exactly what version (latest commit, latest release, latest windows build) or what operating system and with which version of python you are using and what game with what options you are compressing. I will fix this issue as soon I can reproduce it.

**Comment by deadair3210** [comment]
will send the exact version and command line when I get back home

**Comment by deadair3210** [comment]
The command I am using is 

nsz -C "%%F" --level 22 --parseCnmt -V --rm-source

where F is the nsp file that I need converted,  I have it set up in a batch script to compress folders of files one by one. No matter  what file it is, it fails with a not enough memory

**Comment by deadair3210** [comment]
![Annotation 2019-11-26 145402](https://user-images.githubusercontent.com/1053504/69671939-9e793280-105c-11ea-9a8a-1201eeef5d72.jpg)

**Comment by nicoboss** [comment]
Very interesting. Is still wasn't able to reproduce it even with the same game and command on any version (latest commit, pip, portable windows build). What version of python are you using (python --version? I'm on Python Python 3.7.5 and booth 3.6 and 3.7 should be supported while 3.8 should work too but isn't really tested. What nsz build are you even using? Repository download, Pip, or Windows Portable? Does your issue occur on any of these 3 builds? Does it only think your RAM is full or is it full in reality (see Windows preinstalled Task Manager while running nsz)?

Not that it matters for this issue but why are you even using a script and don't give NSZ the folder as input? Booth the NSZ CLI and GUI support folder progressing since its first release.

**Comment by deadair3210** [comment]
Sorry for not giving all the info yesterday, when I responded I was exhausted and obviously forgot to add some things
I'm running Python 3.8, if you want i can try the script in a vm or on linux to see if the problem persists if you think that would help?
I'm using the pip version, not sure where the source for that comes from, whether its a package you build or not

![image](https://user-images.githubusercontent.com/1053504/69744947-ccb34c80-1106-11ea-9566-e8c58ef644e2.png)

**Comment by nicoboss** [comment]
I definitely didn't expect the pip version on Python 3.8 x86. The pip version doesn't even install on 3.8 because of pycryptodome==3.9.0 not being compatible with Python 3.8 as reported in #36. At least for me it definitely refuses to install using "pip install nsz" on latest Python 3.8 x86 as I just tried that inside a clean Windows VM.

Because I already know latest pycryptodome works I downloaded the 2.1.1 source code and removed the dependency from setup.py, installed BuildTools from https://visualstudio.microsoft.com/de/downloads/ and downloaded => Buildtools for Visual Studio 2019 and selected C++ build tools while installing and executed the following commands:
pip install wheel setuptools
python setup.py sdist bdist_wheel
pip install .
I then copied over my .switch folder containing my latest prod.keys and run
"nsz -C "SNES_Online_01008D300C50C000_[v0].nsp" --level 22 --parseCnmt -V --rm-source"
So even on Python 3.8 x86n it worked fine on a clean VM without your issue occurring.

What you should test next:
- Does nsz only think your RAM is full or is it full in reality (see Windows preinstalled Task Manager while running nsz)?
- Try if the issue also occurs on https://github.com/nicoboss/nsz/releases/download/2.1.1/nsz_v2.1.1_win64_portable.zip too
- Uninstall ALL your python versions an install either Python 3.8 using that manual setup described above or just get Python 3.7.5 and use "pip install nsz" (and yes that one comes from me and is equivalent to the latest release). Don't use x86 python as it just makes everything slower for no reason if you are on an x64 system.
- Try a clean VM either Linux or Windows which definitely should solve your issue.

I take such compatibility issues very serious as they might affect a large number of users and am quite happy about all information, I can get to track it down. Thanks a lot for helping me to get this issue solved.

**Comment by deadair3210** [comment]
Not a problem, I enjoy finding issues and helping people fix them. Actually been trying to learn programming for a while now passively so its good practice. I think I'm going to make a clean vm of windows 10 and see if I can reproduce it there. Will let you know when I know more

**Comment by deadair3210** [comment]
Good news! Seems 3.8 was the issue, as replacing 3.8 with 3.75 and reinstalling nsz works perfectly! Thanks! Need any more tests?

**Comment by nicoboss** [comment]
Great to hear that it works fine on 3.75 however it should work on 3.8 too. It would be cool if you could try 3.8 again and see if it works after pip installing it manually. To this this just download https://github.com/nicoboss/nsz/archive/2.1.1.zip and change pycryptodome==3.9.0 to pycryptodome>=3.9.0 inside setup.py. From there just follow the description in my last message.

**Comment by deadair3210** [comment]
Will do when I next get a chance, and will let you know

**Comment by deadair3210** [comment]
Seems to work fine! Oddly,didnt have to manually install it, just pip install nsz and it automatically pulled pycryptodome 3.9
![image](https://user-images.githubusercontent.com/1053504/69828808-fa6dc380-11e2-11ea-8b5c-496583a289a4.png)

**Comment by deadair3210** [comment]
Maybe it had something to do with 32 python on 64bit?

**Comment by nicoboss** [comment]
I finally figured it out. On Python 3.8 and later the user must have Buildtools for Visual Studio 2019 installed or zstandard and pycryptodome==3.9.0 will fail to install as there’s no precompiled python 3.8 version available for them. I tested Python 3.8 x64, Python 3.9a1 x64, Python 3.8 x86 and Python 3.9 x86 and they all worked perfectly fine after installing Buildtools! Not really sure how you got the original issue but I don't think it matters anymore as after properly installing it, nsz works fine on Python 3.8 for you. I will see if there's a way for me to ship precompiled dependencies for all versions and if that isn't possible, I will update the README.md to let users know Buildtools is required for Python 3.8 and above.

**Comment by deadair3210** [comment]
Awesome, yeah its been working fine all day! It refused to install without the buildtools installed the first time, maybe something in the install gets messed up when retried after the tools are installed?
